### PR TITLE
PIM-3076 : re-work the variant group import

### PR DESCRIPTION
**Pull request by nidup**
Re-work the import to be able to create / update variant groups and variant groups values 

In previous PRs
 - [x] Extract / Merge AbstractProcessor on Master

In this PR,
 - [x] update the import of variant groups to create / update variant groups

In coming PRs,
 - update the import of groups to create / update only groups (not VG) WIP
 - update the export of groups to remove the axis column and export only groups (not VG) Another PR

| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | Y
| BC breaks?           | N
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | N
| Fixed tickets        | PIM-3076
| DB schema updated?   |
| Migration script?    |
| Doc PR               |

### CODE REVIEW

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+     * if it does not exist.
+     *
+     * @param ReferableEntityRepositoryInterface $repository the repository to search inside
+     * @param array                              $data       the data that is currently processed
+     * @param string                             $class      the class to instanciate in case the
+     *                                                       object has not been found
+     *
+     * @return object
+     *
+     * @throws MissingIdentifierException in case the processed data do not allow to retrieve an object
```
Comment (fitn) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php]: Does not throw anything

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+     *                                    by its identifiers properly
+     */
+    protected function findOrCreateObject(ReferableEntityRepositoryInterface $repository, array $data, $class)
+    {
+        if (null !== $object = $this->findObject($repository, $data)) {
+            return $object;
+        }
+
+        return new $class();
+    }
```
Comment (fitn) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php]: Why don't you put this kind of method in a manager? I mean it's usefull in controllers and denormalizers. But here you can't manage with factories for example.
Maybe there is too much work for that..?

File: src/Pim/Bundle/BaseConnectorBundle/Writer/Doctrine/VariantGroupWriter.php [hunk 3]
```diff
@@ -145,11 +145,15 @@ protected function copyValuesToProducts(GroupInterface $variantGroup)
     }
 
     /**
-     *
+     * @param GroupInterface $group
      */
-    protected function incrementUpdatedVariantGroupCount()
+    protected function incrementUpdatedVariantGroupCount(GroupInterface $group)
```
Comment (fitn) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/Doctrine/VariantGroupWriter.php]: Maybe this kind of method should be in a DoctrineWriter?

File: src/Pim/Bundle/CatalogBundle/Validator/Constraints/VariantGroupValuesValidator.php [hunk 4]
```diff
@@ -49,7 +49,9 @@ public function validate($variantGroup, Constraint $constraint)
     protected function validateProductValues(GroupInterface $variantGroup, Constraint $constraint)
     {
         $template = $variantGroup->getProductTemplate();
-        $forbiddenAttributes = $variantGroup->getAttributes()->toArray();
+        $forbiddenAttributes = is_array($variantGroup->getAttributes()) ?
+            $variantGroup->getAttributes() :
+            $variantGroup->getAttributes()->toArray();
```
Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Validator/Constraints/VariantGroupValuesValidator.php]: Seems strange. Why don't you have a Collection every time?

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php [hunk 5]
```diff
@@ -11,11 +16,119 @@
...
+    }
+
+    /**
+     * @param GroupInterface $group
+     * @param array          $data
+     */
+    protected function setAxis(GroupInterface $group, $data)
+    {
+        if (isset($data['axis'])) {
+            if (trim($data['axis']) == "") {
```
Comment (fitn) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php]: "" === trim($data['axis']) ?

File: src/Pim/Bundle/CatalogBundle/Validator/Constraints/VariantGroupValuesValidator.php (hunk 4 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Validator/Constraints/VariantGroupValuesValidator.php]: Not on new variant groups

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php [hunk 6]
```diff
@@ -11,11 +16,119 @@
...
+                throw new \LogicException(
+                    sprintf('Axis must be provided for the new variant group "%s"', $data['code'])
+                );
```
Comment (fitn) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php]: Can be the job of the validator? maybe if (isset($data['axis']) && "" === trim...) in the first if can respect more the concerns of the denormalizer?

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php [hunk 7]
```diff
@@ -11,11 +16,119 @@
...
+            $isLabel = false !== strpos($field, 'label-', 0);
+            if ($isLabel) {
+                $labelTokens = explode('-', $field);
+                $localeCode = $labelTokens[1];
+                /** @var GroupTranslation $translation */
+                $translation = $group->getTranslation($localeCode);
+                $translation->setLabel($value);
+                $group->addTranslation($translation);
+            }
+        }
```
Comment (fitn) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php]: TranslatableDenormalizer? Maybe it's really hard to do

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/VariantGroupValuesDenormalizer.php [hunk 8]
```diff
@@ -56,31 +58,35 @@ public function denormalize($data, $class, $format = null, array $context = arra
...
+                    $valueKey .= '-' . $attributeInfos['scope_code'];
+                }
+                if (isset($productValues[$valueKey])) {
+                    $value = $productValues[$valueKey];
+                } else {
+                    $value = new $this->valueClass();
+                    $value->setAttribute($attribute);
+                    $value->setLocale($attributeInfos['locale_code']);
+                    $value->setScope($attributeInfos['scope_code']);
+                }
```
Comment (fitn) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/VariantGroupValuesDenormalizer.php]: Should be in a builder or something like that. I think we already have methods somewhere for that.

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/Flat/VariantGroupProcessor.php [hunk 9]
```diff
@@ -1,46 +1,41 @@
...
  *
  * @author    Nicolas Dupont <nicolas@akeneo.com>
- * @copyright 2014 Akeneo SAS (http://www.akeneo.com)
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
-class VariantGroupValuesProcessor extends AbstractConfigurableStepElement implements
-    ItemProcessorInterface,
-    StepExecutionAwareInterface
+class VariantGroupProcessor extends AbstractProcessor
```
Comment (jjanvier) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/Flat/VariantGroupProcessor.php]: bad naming ^^

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php (hunk 2 above)
Comment (fitn) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php]: In fact, it does not change anything for factories.

File: src/Pim/Bundle/BaseConnectorBundle/Writer/Doctrine/VariantGroupWriter.php (hunk 3 above)
Comment (nidup) [inline src/Pim/Bundle/BaseConnectorBundle/Writer/Doctrine/VariantGroupWriter.php]: Yep, we could, we'll see with coming import re-wamping if we introduce or not

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php (hunk 5 above)
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php]: only if left part is a variable

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php (hunk 7 above)
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php]: Not that hard but i would have a second concret implem before to extract it (coming soon)

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/VariantGroupValuesDenormalizer.php (hunk 8 above)
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/VariantGroupValuesDenormalizer.php]: yes, this part have to be re-worked, especially the construction of the field key

File: src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php (hunk 6 above)
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Denormalizer/Flat/GroupDenormalizer.php]: good catch ! Yes responsibility of the validator !

**Pull request merged**